rust/kernel/types.rs

```diff
@@ -91,3 +91,33 @@ impl<T: PointerWrapper + Deref> PointerWrapper for Pin<T> {
         Pin::new_unchecked(T::from_pointer(p))
     }
 }
+
+/// Runs a cleanup function/closure when dropped.
+///
+/// The [`Cleaner::dismiss`] function prevents the cleanup function from running.
+pub struct Cleaner<T: FnOnce()> {
+    cleanup_func: Option<T>,
```

Comment:
This could be `ManuallyDrop<T>`? On `dismiss` we can call `drop`, and on `drop` we can call `take`. This avoids extra memory.

Reply:
Yes, it could. But the size of the object is 1 byte, which I expect to be optimised out on optimised builds anyway.

I don't think adding `ManuallyDrop` and two extra unsafe calls is worth the loss is readability here.

Reply:
It's indeed usually optimised out, but having a known-to-be-true if statement in the drop method still feels awkward.

(BTW it's `align_of::<T>` extra bytes)

Reply:
> It's indeed usually optimised out, but having a known-to-be-true if statement in the drop method still feels awkward.

The `if` statement isn't always true, it is false when `dismiss` is called.

Reply:
> The `if` statement isn't always true, it is false when `dismiss` is called.

Haha, I know you're going to say this :) I only realized that after I send out that message. Obviously you can add a `forget` and it'll be always true :)

`Option` it's not zero-cost if used in `async fn`. We're unlikely to use `async` any time soon though, so I don't feel too strongly in favor of `ManuallyDrop`. I do prefer to rely as few on the optimiser as possible though.

Also, please add `#[inline]` for all of `new, dismiss, drop`. This helps in opt-level=1.

Reply:
> > The `if` statement isn't always true, it is false when `dismiss` is called.
> 
> Haha, I know you're going to say this :) I only realized that after I send out that message. Obviously you can add a `forget` and it'll be always true :)

:)

> `Option` it's not zero-cost if used in `async fn`. We're unlikely to use `async` any time soon though, so I don't feel too strongly in favor of `ManuallyDrop`. I do prefer to rely as few on the optimiser as possible though.

I actually liked your idea of using `ManuallyDrop` and was making the changes until I got compilation errors because `drop` and `take` were unsafe -- that's what tipped the scales against it for me.

> Also, please add `#[inline]` for all of `new, dismiss, drop`. This helps in opt-level<=1.

I'd rather not add unnecessary pollution. For optimisation levels <= 1, we obviously don't care about performance as they are meant for debugging, and in such cases I'd rather not have functions inlined so that the asm instructions are as similar as possible to the source code.

Reply:
It's quite common pattern to have `#[inline]` on them; std libraries always have `#[inline]` for those short, not-so-interesting functions, even if they are generic.

One common issue with Rust debugging is that there are too many intermediate function frames before you reach your function-of-interest: in this case the debugger clearly does not care about the `take` etc in `ScopeGuard`, but rather the actually function getting called, having them inlined help people navigate to the part they actually care about.

Reply:
I would agree with Gary in most cases. However, given the lesser optimization levels we provide are just a "bonus", I agree with Wedson here that adding clutter just for those levels is not advisable. We may even need to drop those levels (we will see what happens), so... :(

If a particular hint helps on `-Copt-level=2`, though, that would be different.

Related: the `str.rs` PR had a bunch of `#[inline]` -- if those were only for debugging purposes, I would remove them for the time being.

Reply:
For CStr they are necessary -- CStr does not have generics, so they are not inlineable without `#[inline]`.

Reply:
One argument would be that it might be more straightforward to require `#[inline]` to be added to all short and simple functions. Otherwise we need to say "add `#[inline]` if it's a short and simple function, and does not depend on any generics; don't add it if generics are used".

After all, adding an extra one is just an unnecessary hint, forgetting one will be some significant performance loss.

Reply:
You mean across crates? I think most production kernels should be using LTO.

Having said that, the kernel is quite keen in general adding `inline` hints everywhere, so perhaps people will end up doing it. On the other hand, in general, if we could show similar performance *without* any hints (assuming LTO), it would be a very nice point, because we could say "forget about all that".

Reply:
For non-generics, if no `#[inline]` is added, code has to be generated for the crate that defines it. If `#[inline]` is added then no code is generated.

As for LTO, I am not so sure about how LTO would work with kernel modules; likely some non-inlined code still has to go to vmlinux.

Regardless, we probably don't want to *require* LTO to get reasonable performance. I think it's generally a good practice to add `#[inline]`.

Reply:
The compiler can still figure out most of the functions that could possibly benefit from inlining, and generate generic-like bytecode/IR or whatever it is called instead for those cases.

Obviously, manual hints are better, but the problem with hints like that is that they get very annoying, very soon. And, in Rust, we take up typically an entire line just for that.

If the compiler could remove the need for *most* of them in non-LTO builds, then the hints that need to remain are even more useful for human readers and can be re-inspected as needed to assess performance etc.

As for reasonable performance: if someone does not use LTO, then they do not care *that much* about performance. It would be best to have the data and make comparisons, rather than guessing.

As for kernel modules, I am not sure what you mean. Do you mean out-of-tree loadable kernel modules? Given we use generics, we need to provide them with the metadata anyway.

Reply:
> The compiler can still figure out most of the functions that could possibly benefit from inlining, and generate generic-like bytecode/IR (not sure how it is called in rustc) instead for those cases.

Well, Rust does not guess whether something is suitable for inlining at all; it's all up to LLVM. Therefore for non-generic functions, cross-crate inlining requires either `#[inline]` or LTO. It's like `static inline` function vs an `extern` function.

> As for reasonable performance: if someone does not use LTO, then they do not care that much about performance. It would be best to have the data and make comparisons, rather than guessing.

They may not care *that much*, but requiring a function call for `&CStr` to `*const c_char` conversion is not acceptable.

To be clear, I just say that we should have `#[inline]` for non-generic functions; for generic ones it's okay either way, but I do prefer consistency.

Reply:
Also LTO probably means that we would need LLD somewhere in our pipeline? Anyway I still think we shouldn't require or assume people will use LTO.

Reply:
By the way, please do not take me the wrong way when I complain about stuff like this. I usually push you guys a lot on (@nbdd0121 and @bjorn3) because I know you are experts on this stuff.

The idea being that either 1) you tell me how I am completely stupid (and then I learn more about `rustc`/LLVM internals, which is a win for me; plus we collectively learn how to answer similar questions if they happen to be raised in e.g. the LKML); or 2) I convince you to improve Rust/LLVM in some areas and everybody wins ;-)

Reply:
> Well, Rust does not guess whether something is suitable for inlining at all; it's all up to LLVM. Therefore for non-generic functions, cross-crate inlining requires either `#[inline]` or LTO. It's like `static inline` function vs an `extern` function.

I was thinking LLVM could give back its cost heuristics after optimizing a function, and then the frontend decide if it wants to generate a Rust generic-like function or go with a normal one etc.

Yeah, it is likely not trivial if LLVM is not setup to do that, and most likely not worth it anyway since using LTO is simpler and a more general approach. But hey, it would improve things for all non-LTO-enabled projects *and* remove the needs for so many hints.

> They may not care _that much_, but requiring a function call for `&CStr` to `*const c_char` conversion is not acceptable.

Yes, but what I am saying is the amount of hints we use depends on whether we say LTO is the way to go or not, and how much the overall % lost. The idea being that we could try to be better than C here and have the hints that actually matter, instead of systematically providing hints everywhere.

At the current time, the safest approach is indeed doing the same the C side does (which is providing hints, having functions in headers, using macros to hack it, etc.).

Reply:
> Also LTO probably means that we would need LLD somewhere in our pipeline? Anyway I still think we shouldn't require or assume people will use LTO.

Yeah:

```
config HAS_LTO_CLANG
	def_bool y
	# Clang >= 11: https://github.com/ClangBuiltLinux/linux/issues/510
	depends on CC_IS_CLANG && CLANG_VERSION >= 110000 && LD_IS_LLD
```

It is still experimental, but since it is also a requirement for CFI, I bet it will become way more common soon enough.

Reply:
> Also LTO probably means that we would need LLD somewhere in our pipeline? Anyway I still think we shouldn't require or assume people will use LTO.

Rustc's LTO support normally doesn't require LLD as rustc performs it itself. Due to the way rustc is used, it can't do LTO itself though and will need linker plugin LTO. This should work with linkers other than LLD though, but requires a linker plugin for the right LLVM version rustc uses.

> I was thinking LLVM could give back its cost heuristics after optimizing a function, and then the frontend decide if it wants to generate a Rust generic-like function or go with a normal one etc.

By the time codegen happens, the frontend has already emitted the metadata file.